### PR TITLE
tools/scylla-sstable introduce the write operation

### PR DESCRIPTION
**Pull request by denesb**
Implementing json2sstable functionality. It allows generating an sstable from a JSON description of its content. Uses identical schema to dump-data, so it is possible to regenerate an existing sstable, by feeding the output of dump-data to write.
Most of the scylla storage engine features are supported. The only non-supported features are counters and non-strictly atomic data types (including frozen collections, tuples and UDTs).

Example invocation:
```
scylla sstable write --system-schema system_schema.columns --input-file ./input.json --generation 0
```
    
Refs: https://github.com/scylladb/scylladb/issues/9681

Future plans:
* Complete support for remaining features (counters and non-atomic types).
* Make sstable format configurable on the command line.

### CODE REVIEW

**Comment by bhalevy** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1622/

@denesb 

```
[2022-08-01T12:25:04.316Z] FAILED: build/debug/tools/scylla-sstable.o 
[2022-08-01T12:25:04.316Z] clang++ -MD -MT build/debug/tools/scylla-sstable.o -MF build/debug/tools/scylla-sstable.o.d -I/jenkins/workspace/releng/Scylla-CI/scylla/seastar/include -I/jenkins/workspace/releng/Scylla-CI/scylla/build/debug/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -DSEASTAR_API_LEVEL=6 -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_TYPE_ERASE_MORE -DFMT_LOCALE -DFMT_SHARED -I/usr/include/p11-kit-1  -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -Og -DSCYLLA_ENABLE_WASMTIME -g -gz -iquote. -iquote build/debug/gen --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/releng/Scylla-CI/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK   -Iabseil -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-narrowing -Wno-array-bounds -Wno-nonnull -Wno-uninitialized -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/debug/tools/scylla-sstable.o tools/scylla-sstable.cc
[2022-08-01T12:25:04.316Z] tools/scylla-sstable.cc:1612:49: error: no member named 'runtime' in namespace 'fmt'
[2022-08-01T12:25:04.316Z]             auto parse_error = fmt::format(fmt::runtime(msg), std::forward<decltype(args)>(args)...);
[2022-08-01T12:25:04.316Z]                                            ~~~~~^
```

**Comment by denesb** [comment]
Yes, I've seen it. On F35 it won't compile without the `fmt::runtime`, in dbuild it won't compile with.

**Comment by bhalevy** [comment]
dbuild trumps local toolchain :)

**Comment by denesb** [comment]
> dbuild trumps local toolchain :)

This will be of no consolation to all the people whose setup I'll break. I ended up using an `#ifdef`.

**Comment by denesb** [comment]
New in v1.1:
* Fixed compile error around `fmt::runtime()` (hopefully).

File: tools/scylla-sstable.cc [hunk 1]
```diff
@@ -869,30 +869,52 @@ void consume_sstables(schema_ptr schema, reader_permit permit, std::vector<sstab
     }
 }
 
-using operation_func = void(*)(schema_ptr, reader_permit, const std::vector<sstables::shared_sstable>&, const bpo::variables_map&);
+using read_operation_func = void(*)(schema_ptr, reader_permit, const std::vector<sstables::shared_sstable>&, const bpo::variables_map&);
+using write_operation_func = void(*)(schema_ptr, reader_permit, sstables::sstables_manager&, const bpo::variables_map&);
```
Comment (avikivity) [inline tools/scylla-sstable.cc:875]: I can see operations that do both reads and writes (`compact` operation to compact a bunch of sstables). So how about extending operation_func with sstables_manager? pure writes will have empty input sstables vector, and pure reads will ignore the sstables manager, and a bunch of checks below will disappear.

File: tools/scylla-sstable.cc [hunk 2]
```diff
@@ -2137,11 +2158,12 @@ Validate the specified sstables:
...
-    app.add_positional_options({
-        {"sstables", bpo::value<std::vector<sstring>>(), "sstable(s) to process, can also be provided as positional arguments", -1},
-    });
-
     if (found_op) {
+        if (found_op->kind() == operation::kind::read) {
+            app.add_positional_options({
+                {"sstables", bpo::value<std::vector<sstring>>(), "sstable(s) to process, can also be provided as positional arguments", -1},
+            });
+        }
```
Comment (avikivity) [inline tools/scylla-sstable.cc:2856]: This runtime construction kind of runs --help. I don't see a way to fix it with boost program_options though, it's too outdated. Python has great support for it via add_subparsers().

Comment (denesb) [reply tools/scylla-sstable.cc:2856]: Options added this way still appear in the `--help` output. In fact that is the entire point in playing these games here: I want only the options relevant to a given command to appear in its `--help` output.

Comment (denesb) [reply tools/scylla-sstable.cc:2856]: You are right that these won't appear in `scylla sstable --help`, but they will in `scylla sstable dump-data --help`, an operation to which the `sstables` option is relevant.

**Comment by avikivity** [comment]
Looks good.

File: tools/scylla-sstable.cc (hunk 2 above)
Comment (avikivity) [reply tools/scylla-sstable.cc:2856]: Ah, so it's not too bad. In any case we can't do much about it.

File: tools/scylla-sstable.cc (hunk 1 above)
Comment (denesb) [reply tools/scylla-sstable.cc:875]: Good idea, will do that.

**Comment by denesb** [comment]
New in v2:
* Completed support for most of the basic scylla storage engine features. The only unsupported features are now counters and non-strictly atomic data types (including frozen collections, tuples and UDTs).
* Improved diagnostics on parse failure: error now contains line number + offset.
* Operations are now all read-write, instead of distinct read and write operations.
* Much improved tests: improve schema and data used to cover all the supported features.
* Test accidentally compared the golden-dump to itself :facepalm:. This is now fixed, the dump-under-test is now compared to the golden-dump (the test first dumps an sstable to json, then converts the json to an sstable, then dumps it again, comparing the two dumps at the end).

**Comment by avikivity** [comment]
> New in v2:
> 
> * Completed support for most of the basic scylla storage engine features. The only unsupported features are now counters and non-strictly atomic data types (including frozen collections, tuples and UDTs).
> * Improved diagnostics on parse failure: error now contains line number + offset.
> * Operations are now all read-write, instead of distinct read and write operations.
> * Much improved tests: improve schema and data used to cover all the supported features.
> * Test accidentally compared the golden-dump to itself facepalm. This is now fixed, the dump-under-test is now compared to the golden-dump (the test first dumps an sstable to json, then converts the json to an sstable, then dumps it again, comparing the two dumps at the end).

Please update the cover letter. Also include an example invocation line, in case it makes it into the newsletter.

**Comment by denesb** [comment]
> Please update the cover letter. Also include an example invocation line, in case it makes it into the newsletter.

Done.

**Comment by denesb** [comment]
One item forgotten from the v2 changelog: sstable write validation is now set to the strictest level, but this is configurable with the new `--validation-level` flag.

**Comment by bhalevy** [comment]
> One item forgotten from the v2 changelog: sstable write validation is now set to the strictest level, but this is configurable with the new `--validation-level` flag.

In the tool only, not in the scylla server, right?

**Comment by denesb** [comment]
> > One item forgotten from the v2 changelog: sstable write validation is now set to the strictest level, but this is configurable with the new `--validation-level` flag.
> 
> In the tool only, not in the scylla server, right?

Yes of course. The validator is configurable on a per-sstable-write level.

**Pull request closed without merging**